gym/tests/test_core.py

```diff
@@ -9,9 +12,99 @@ def __init__(self, arg):
         self.arg = arg
 
 
+class UnittestEnv(core.Env):
+    observation_space = spaces.Box(low=0, high=255, shape=(64, 64, 3), dtype=np.uint8)
+    action_space = spaces.Discrete(3)
+
+    def reset(self):
+        return self.observation_space.sample()  # Dummy observation
+
+    def step(self, action):
+        observation = self.observation_space.sample()  # Dummy observation
+        return (observation, 0.0, False, {})
+
+
+class UnknownSpacesEnv(core.Env):
+    """This environment defines its observation & action spaces only
+    after the first call to reset. Although this pattern is sometimes
+    necessary when implementing a new environment (e.g. if it depends
+    on external resources), it is not encouraged.
+    """
+
+    def reset(self):
+        self.observation_space = spaces.Box(
+            low=0, high=255, shape=(64, 64, 3), dtype=np.uint8
+        )
+        self.action_space = spaces.Discrete(3)
+        return self.observation_space.sample()  # Dummy observation
+
+    def step(self, action):
+        observation = self.observation_space.sample()  # Dummy observation
+        return (observation, 0.0, False, {})
+
+
+class NewPropertyWrapper(core.Wrapper):
+    def __init__(
+        self,
+        env,
+        observation_space=None,
+        action_space=None,
+        reward_range=None,
+        metadata=None,
+    ):
+        super().__init__(env)
+        if observation_space is not None:
+            # Only set the observation space if not None to test property forwarding
+            self.observation_space = observation_space
+        if action_space is not None:
+            self.action_space = action_space
+        if reward_range is not None:
+            self.reward_range = reward_range
+        if metadata is not None:
+            self.metadata = metadata
+
+
 def test_env_instantiation():
     # This looks like a pretty trivial, but given our usage of
     # __new__, it's worth having.
     env = ArgumentEnv("arg")
     assert env.arg == "arg"
     assert env.calls == 1
+
+
+properties = [
+    {
+        "observation_space": spaces.Box(
+            low=0.0, high=1.0, shape=(64, 64, 3), dtype=np.float32
+        )
+    },
+    {"action_space": spaces.Discrete(2)},
+    {"reward_range": (-1.0, 1.0)},
+    {"metadata": {"render.modes": ["human", "rgb_array"]}},
+    {
+        "observation_space": spaces.Box(
+            low=0.0, high=1.0, shape=(64, 64, 3), dtype=np.float32
+        ),
+        "action_space": spaces.Discrete(2),
+    },
+]
+
+
+@pytest.mark.parametrize("klass", [UnittestEnv, UnknownSpacesEnv])
```

Comment:
Minor nitpick - is `klass` often used instead of `class`? I'd intuitively go rather for `cls`

Reply:
By convention, [`cls` is reserved for classmethods](https://www.python.org/dev/peps/pep-0008/#function-and-method-arguments). PEP8 suggests using `class_` instead, but `klass` is also a fairly common name, even [CPython uses it](https://github.com/python/cpython/search?q=klass).